RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc

```diff
@@ -0,0 +1,474 @@
+GEMSegmentAlgorithm::ProtoSegments 
+GEMSegmentAlgorithm::chainHits(const EnsembleHitContainer& rechits) {
+
+  ProtoSegments rechits_chains; 
+  EnsembleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector <bool> usedCluster;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
+    usedCluster.push_back(false);
+  }
+
+  // merge chains that are too close ("touch" each other)
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    for(size_t MMM = NNN+1; MMM < seeds.size(); ++MMM) {
+      if(usedCluster[MMM] || usedCluster[NNN]){
+        continue;
+      }
+      // all is in the way we define "good";
+      // try not to "cluster" the hits but to "chain" them;
+      // it does the clustering but also does a better job
+      // for inclined tracks (not clustering them together;
+      // crossed tracks would be still clustered together) 
+      // 22.12.09: In fact it is not much more different 
+      // than the "clustering", we just introduce another
+      // variable in the game - Z. And it makes sense 
+      // to re-introduce Y (or actually wire group mumber)
+      // in a similar way as for the strip number - see
+      // the code below.
+      bool goodToMerge  = isGoodToMerge(seeds[NNN], seeds[MMM]);
+      if(goodToMerge){
+        // merge chains!
+        // merge by adding seed NNN to seed MMM and erasing seed NNN
+
+        // add seed NNN to MMM (lower to larger number)
+        seeds[MMM].insert(seeds[MMM].end(),seeds[NNN].begin(),seeds[NNN].end());
+
+        // mark seed NNN as used
+        usedCluster[NNN] = true;
+        // we have merged a seed (NNN) to the highter seed (MMM) - need to contimue to
+        // next seed (NNN+1)
+        break;
+      }
+
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with
+  // erasing the merged ones, rather than the
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(usedCluster[NNN]) continue; //skip seeds that have been marked as used up in merging
+    rechits_chains.push_back(seeds[NNN]);
+  }
+
+  //***************************************************************
+
+  return rechits_chains;
+}
+
+bool GEMSegmentAlgorithm::isGoodToMerge(const EnsembleHitContainer& newChain, const EnsembleHitContainer& oldChain) {
+
+  std::vector<float> phi_new, eta_new, phi_old, eta_old;
+  phi_new.reserve(newChain.size()); phi_old.reserve(oldChain.size());
+  eta_new.reserve(newChain.size()); eta_old.reserve(oldChain.size());
+  std::vector<int> layer_new, layer_old;
+  layer_new.reserve(newChain.size()); layer_old.reserve(oldChain.size());
+  std::vector<int> bx_new, bx_old;
+  bx_new.reserve(newChain.size()); bx_old.reserve(oldChain.size());
+
+  for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
```

Comment:
I belatedly realize that all of the "new" variable calculations could be delayed until the loop starting on 342 (which would mean only the ones used before the function exits would be computed.